src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfo.cs

```diff
@@ -469,7 +489,20 @@ internal JsonTypeInfo JsonTypeInfo
         {
             get
             {
-                return _jsonTypeInfo ??= Options.GetOrAddJsonTypeInfo(PropertyType);
+                if (_jsonTypeInfo != null)
+                {
+                    // We should not call it on set as it's usually called during initialization
```

Comment:
We could make sure that responsibility for this lives in the `GetOrAddJsonInfo` implementation (so it only happens once, and don't have to recheck in case of a cache hit).

Reply:
it does ensure that already. The reason it's also here is that source gen assigns `_jsonTypeInfo` for some properties during construction and those are not configured in such case so we need to ensure both cases are covered. The reason we can't do it during construction is that it could cause StackOverflow if there is cycle in the object graph